app/src/test/java/ch/epfl/favo/common/TaskToFutureAdapterTest.java

```diff
@@ -0,0 +1,97 @@
+package ch.epfl.favo.common;
+
+import android.app.Activity;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+
+import org.junit.Test;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+public class TaskToFutureAdapterTest {
+
+  @Test
+  public void addTwoNumbersTask() {
+    Task<Void> task = new Task<Void>() {
+      @Override
+      public boolean isComplete() {
+        return false;
+      }
+
+      @Override
+      public boolean isSuccessful() {
+        return false;
+      }
+
+      @Override
+      public boolean isCanceled() {
+        return false;
+      }
+
+      @Nullable
+      @Override
+      public Void getResult() {
+        return null;
+      }
+
+      @Nullable
+      @Override
+      public <X extends Throwable> Void getResult(@NonNull Class<X> aClass) throws X {
+        return null;
+      }
+
+      @Nullable
+      @Override
+      public Exception getException() {
+        return null;
+      }
+
+      @NonNull
+      @Override
+      public Task<Void> addOnSuccessListener(@NonNull OnSuccessListener<? super Void> onSuccessListener) {
+        return null;
+      }
+
+      @NonNull
+      @Override
+      public Task<Void> addOnSuccessListener(@NonNull Executor executor, @NonNull OnSuccessListener<? super Void> onSuccessListener) {
+        return null;
+      }
+
+      @NonNull
+      @Override
+      public Task<Void> addOnSuccessListener(@NonNull Activity activity, @NonNull OnSuccessListener<? super Void> onSuccessListener) {
+        return null;
+      }
+
+      @NonNull
+      @Override
+      public Task<Void> addOnFailureListener(@NonNull OnFailureListener onFailureListener) {
+        return null;
+      }
+
+      @NonNull
+      @Override
+      public Task<Void> addOnFailureListener(@NonNull Executor executor, @NonNull OnFailureListener onFailureListener) {
+        return null;
+      }
+
+      @NonNull
+      @Override
+      public Task<Void> addOnFailureListener(@NonNull Activity activity, @NonNull OnFailureListener onFailureListener) {
+        return null;
+      }
+    };
+
+    CompletableFuture<Void> future = new TaskToFutureAdapter<Void>(task);
+    future.complete(null);
+    assert(true);
```

Comment:
You are not testing that the adapter actually works (don't do tests that do not actually check things, we will penalize those). You should do the following:

- To your Task override, implement addOnSuccessListener and addOnFailureListener such that they actually register the passed listeners to some internal state (for instance create 2 ArrayLists of callbacks to a member of your Task).
- Also implement a `complete()` function that triggers all registered listeners in your success arraylist of callbacks with some meaningful value, and a `fail()` function that triggers all the ones in the failure arraylist.
- Then, check that it actually runs the callback (in the test **you want to block the promise to make sure that it finishes, otherwise you will not check it**).

Reply:
@arosu here's a working implementation:

In your `src` folder in some package: the adapter

TaskToFutureAdapter.java
```java
public class TaskToFutureAdapter<T> extends CompletableFuture<T> {
    TaskToFutureAdapter(Task<T> task) {
        super();
        task.addOnSuccessListener(this::complete);
        task.addOnFailureListener(this::completeExceptionally);
    }
}
```

In your `test` folder, as an util: this TestTask will allow you to force complete and fail it, to check if the listeners are actually called in your tests.

TestTask.java
```java

class TestTask extends Task<String> {
    // possible task status
    enum TaskStatus {PENDING, SUCCESS, FAILURE, CANCELLED}

    // success listeners
    private final List<OnSuccessListener> successes = new ArrayList<>();
    // failure listeners
    private final List<OnFailureListener> failures = new ArrayList<>();
    // task status
    private TaskStatus status = TaskStatus.PENDING;
    // value of the successful task, constant for testing
    public final String value = "Hello world!";
    // exception to throw if task is not successful
    public final RuntimeException exception = new RuntimeException("Task was not finished successfully");

    @Override
    public boolean isComplete() {
        return status != TaskStatus.PENDING;
    }

    @Override
    public boolean isSuccessful() {
        return status == TaskStatus.SUCCESS;
    }

    @Override
    public boolean isCanceled() {
        return status == TaskStatus.CANCELLED;
    }

    @Nullable
    @Override
    public String getResult() {
        if (status != TaskStatus.SUCCESS) {
            throw exception;
        }
        return value;
    }

    @Nullable
    @Override
    public <X extends Throwable> String getResult(@NonNull Class<X> aClass) throws X {
        if (status != TaskStatus.SUCCESS) {
            throw exception;
        }
        return value;
    }

    @Nullable
    @Override
    public Exception getException() {
        if (status != TaskStatus.FAILURE) {
            throw new RuntimeException("Task did not fail");
        }
        return exception;
    }

    @NonNull
    @Override
    public Task<String> addOnSuccessListener(@NonNull OnSuccessListener<? super String> onSuccessListener) {
        successes.add(onSuccessListener);
        return this;
    }

    @NonNull
    @Override
    public Task<String> addOnSuccessListener(@NonNull Executor executor, @NonNull OnSuccessListener<? super String> onSuccessListener) {
        return addOnSuccessListener(onSuccessListener);
    }

    @NonNull
    @Override
    public Task<String> addOnSuccessListener(@NonNull Activity activity, @NonNull OnSuccessListener<? super String> onSuccessListener) {
        return addOnSuccessListener(onSuccessListener);
    }

    @NonNull
    @Override
    public Task<String> addOnFailureListener(@NonNull OnFailureListener onFailureListener) {
        failures.add(onFailureListener);
        return this;
    }

    @NonNull
    @Override
    public Task<String> addOnFailureListener(@NonNull Executor executor, @NonNull OnFailureListener onFailureListener) {
        return addOnFailureListener(onFailureListener);
    }

    @NonNull
    @Override
    public Task<String> addOnFailureListener(@NonNull Activity activity, @NonNull OnFailureListener onFailureListener) {
        return addOnFailureListener(onFailureListener);
    }

    // force the success of the task
    public void complete() {
        status = TaskStatus.SUCCESS;
        for (final OnSuccessListener success : successes) {
            success.onSuccess(value);
        }
    }

    // force failure of the task
    public void fail() {
        status = TaskStatus.FAILURE;
        for (final OnFailureListener failure : failures) {
            failure.onFailure(exception);
        }
    }
}

```

Test cases in your test suite class: they actually test that the registered listeners get run, and that you force wait the test (which is the only place where you should block a thread to wait for a future result!) to check that it actually finishes as expected.

```java
@Test
    public void taskToFutureAdapterSuccess() throws ExecutionException, InterruptedException {
        TestTask task = new TestTask();
        CompletableFuture<String> f = new TaskToFutureAdapter<>(task);
        task.complete();
        // check that the listener is triggered properly (MAY BE LATER)
        f.thenApply(value -> {
            assertEquals(task.value, value);
            return value;
        });
        // explicitly wait because we want the test to actually check the future termination
        String result = f.get();
        assertEquals(task.value, result);
    }

    @Test
    public void taskToFutureAdapterFailure() throws InterruptedException {
        TestTask task = new TestTask();
        CompletableFuture<String> f = new TaskToFutureAdapter<>(task);
        task.fail();
        // check that listener is triggered properly (MAY BE LATER)
        f.exceptionally(exception -> {
            assertEquals(task.exception, exception);
            return null;
        });
        // explicitly wait because we want the test to actually check the future termination
        try {
            f.get();
        } catch (ExecutionException e) {
            assertEquals(task.exception, e.getCause());
        }
    }
```

Reply:
Wow awesome, thank you so much for this!